ServiceLoader.load() should be in a privileged action

kabir opened the pull request:
Without this when running with a security manager I get exceptions in WildFly

Emily-Jiang commented:
@kabir can you resolve the merge conflict?

Emily-Jiang commented:
@cescoffier Do you foresee any issues to take this change in?

cescoffier commented:
No, it should be fine.

Emily-Jiang commented:
@kabir please solve the merge conflict and then mention this change in the release note of this spec doc so that we can take this PR into the 3.0 release.

kabir commented:
@Emily-Jiang Rebased. In the new version I switched to PrivilegedExceptionAction since the multiple provider check throws an exception which would have been swallowed with what I had originally.

I've added a release not entry. Hopefully, the location and format is acceptable.

Azquelt commented on api/src/main/java/org/eclipse/microprofile/reactive/streams/operators/spi/ReactiveStreamsFactoryResolver.java:
```diff
@@ -77,17 +79,28 @@ private static ReactiveStreamsFactory loadFromSpi(ClassLoader cl) {
+                });
+            } catch (PrivilegedActionException e) {
+                Throwable t = e.getCause();
+                if (t instanceof RuntimeException) {
+                    throw new RuntimeException(t);
                 }
+                throw new RuntimeException(t);
```
At the moment this `if` block does the same thing whether the condition is true or not.
```suggestion
                if (t instanceof RuntimeException) {
                    throw (RuntimeException) t;
                }
                if (t instanceof Error) {
                    throw (Error) t;
                }
                throw new RuntimeException(t);
```

kabir replied:
You're right - I was in a rush :-)

And I never realised until now that Error is unchecked!

Should be fixed now, thanks

Azquelt commented on core/src/main/java/org/eclipse/microprofile/reactive/streams/operators/core/ReactiveStreamsEngineResolver.java:
```diff
@@ -76,17 +78,28 @@ private static ReactiveStreamsEngine loadFromSpi(ClassLoader cl) {
+                });
+            } catch (PrivilegedActionException e) {
+                Throwable t = e.getCause();
+                if (t instanceof RuntimeException) {
+                    throw new RuntimeException(t);
                 }
+                throw new RuntimeException(t);
```
This needs the same change as above.

Emily-Jiang commented on spec/src/main/asciidoc/release_notes.asciidoc:
```diff
@@ -35,11 +35,12 @@ A full list of changes delivered in the 2.0 release can be found at link:https:/
 - Update to Jakarta EE8 APIs for MP 4.0 (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/128[#128])
 - Javadoc clarification: (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/135[#135]) (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/137[#137])(link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/142[#142]) (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/129[#129])
 - TCK clarification: (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/141[#141])
+- Loading ReactiveStreamsEngine and ReactiveStreamsFactory should also work running with a security manager (link:https://github.com/eclipse/microprofile-reactive-streams-operators/pull/170[#170])
```
You should add a heading Other Changes on the release notes 3.0 and put this line 45 over there. Please don't update the previous release notes.

kabir replied:
@Emily-Jiang - is it fine now?

Emily-Jiang commented on api/src/main/java/org/eclipse/microprofile/reactive/streams/operators/spi/ReactiveStreamsFactoryResolver.java:
```diff
@@ -77,17 +79,31 @@ private static ReactiveStreamsFactory loadFromSpi(ClassLoader cl) {
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                Throwable t = e.getCause();
```
Why does the unwrapping? Why not just throw e?

Azquelt replied:
`PrivilegedActionException` is checked so it can't be thrown without wrapping it again.

I guess you could wrap it again in a `RuntimeException` but I think it would be more helpful to the caller to unwrap the `PrivilegedActionException` since that's only created because we called `AccessController.doPrivileged` which is an implementation detail of this method.

kabir replied:
I agree with @Azquelt - this is what we normally do in the WildFly codebase

Emily-Jiang commented on spec/src/main/asciidoc/release_notes.asciidoc:
```diff
@@ -35,11 +38,12 @@ A full list of changes delivered in the 2.0 release can be found at link:https:/
 - ReactiveStreamsFactory.fromSubscriber parameter type change (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/134[#134])
 
 === Other Changes
 - Update to Jakarta EE8 APIs for MP 4.0 (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/128[#128])
```
This section should not be deleted.

kabir replied:
This is not deleted :-) The '-' is part of the text to make a list

Emily-Jiang commented on spec/src/main/asciidoc/release_notes.asciidoc:
```diff
@@ -35,11 +38,12 @@ A full list of changes delivered in the 2.0 release can be found at link:https:/
 
 - OSGi dependencies marked as "provided" (link:https://github.com/eclipse/microprofile-reactive-streams-operators/issues/132[#132])
```
Please don't change the previous release notes. All of the updates for the release notes on 2.0 need to be reverted.

kabir replied:
Not deleted, '-' is for a list
